spec/options.spec.js

```diff
@@ -43,43 +52,39 @@ describe('Ajv Options', function () {
       object.should.have.property('foo');
       object.should.have.property('bar');
       object.should.not.have.property('baz');
+
+      var failingObject = {
+        foo: 'foo', bar: 42, baz: 'baz-to-be-removed'
+      };
+
+      ajv.validate('//test/fooBar', failingObject).should.equal(false);
+      failingObject.should.have.property('foo');
+      failingObject.should.have.property('bar');
+      failingObject.should.not.have.property('baz');
     });
 
 
     it('should remove properties that would error when `additionalProperties` is a schema', function() {
       var ajv = new Ajv({ removeAdditional: 'failing' });
 
       ajv.addSchema({
-        id: '//test/fooBar',
-        properties: { foo: { type: 'string' }, bar: { type: 'string' } },
-        additionalProperties: { type: 'string' }
-      });
-
-      var object = {
-        foo: 'foo', bar: 'bar', baz: 'baz-to-be-kept', fizz: 1000
-      };
-
-      ajv.validate('//test/fooBar', object).should.equal(true);
-      object.should.have.property('foo');
-      object.should.have.property('bar');
-      object.should.have.property('baz');
-      object.should.not.have.property('fizz');
-
-      ajv.addSchema({
         id: '//test/fooBar2',
         properties: { foo: { type: 'string' }, bar: { type: 'string' } },
         additionalProperties: { type: 'string', pattern: '^to-be-', maxLength: 10 }
       });
 
-      object = {
-        foo: 'foo', bar: 'bar', baz: 'to-be-kept', quux: 'to-be-removed', fizz: 1000
+      var failingObject = {
+        foo: 232,
+        bar: 42,
+        fizz: 'buzz'
       };
 
-      ajv.validate('//test/fooBar2', object).should.equal(true);
-      object.should.have.property('foo');
-      object.should.have.property('bar');
-      object.should.have.property('baz');
-      object.should.not.have.property('fizz');
+      console.log(ajv.getSchema('//test/fooBar2').toString());
+      ajv.validate('//test/fooBar2', failingObject).should.equal(false);
```

Comment:
@epoberezkin I think I found a bug in master. When `additionalProperties` is a schema, properties are not validate (and this test fails at this location 😬 .

Reply:
(it's only a problem if you supply an invalid additional property with an additional properties schema, and `removeAdditional: 'failing' )

Reply:
@epoberezkin I figured out the issue. `$nextValid` was not being unset when the additional property was removed, which caused the function to exit with no errors.